Fix blackboard content link

LMS007 opened the pull request:
Make the no file message href dynamic 

Change the LMSFilePicker to render a different documentation help link when no files are returned for Blackboard vs Canvas.

------

Update

I have refactored this PR to no longer use a provider and instead take a `missingFilesHelpLink` prop from ContentSelector.
Secondly, we have decided the 0 way to link to Blackboard docs is to link to an internal alias we maintain that has external links for file upload documentation for both ultra and classic version so blackboard. Currently there is no way to differentiate between these versions from the perspective of the client or config.

Slack discussion here
https://hypothes-is.slack.com/archives/C2BLQDKHA/p1625075688421600

Fixes https://github.com/hypothesis/lms/issues/2860

LMS007 commented on lms/static/scripts/frontend_apps/components/test/LMSFilePicker-test.js:
```diff
@@ -46,6 +62,9 @@ describe('LMSFilePicker', () => {
       './Dialog': FakeDialog,
+      // Don't mock <FileList> because <NoFiles> requires
+      // it to render for code coverage.
+      './FileList': FileList,
```
Some hackery here, because `<NoFiles>` gets passed as a prop, it won't render in this tests (e.g. not full codecov) unless we don't mock `<FileList>`, so this simply forced FileList to render too. 

I also felt `NoFiles` should **continue** being passed as prop and not a child component because it's not a central child component of `FileLIst`.

robertknight reviewed (changes requested):
The design on master is that `LMSFilePicker` is agnostic of LMS-specific details and app configuration, which are handled by the parent `ContentSelector`. The current status on this branch is that `LMSFilePicker` handles some of the LMS-specific details (the help link) whereas the parent `ContentSelector` handles other details (which API method to call to list the files). This is an inconsistency that we should avoid.

If the UI of the dialog needs to diverge significantly depending on the LMS then it will probably make sense to change the responsibilities of the components. However for customizing a single help link I think it would be better to extend the current design where LMS-specific knowledge remains in `ContentSelector`. For example the LMS-specific links can be passed in as a prop to `LMSFilePicker` by `ContentSelector`, which already has different branches that render `LMSFilePicker` with props that depend on the LMS.

LMS007 commented:
Updated PR:

Rob, good suggestions about the provider and just using ContentSelector instead, that simplifies the code a bit. 

I also changed the bb help link to an internal WP alias that we'll maintain. Nate & Matt are working on adding it. I'll probably hold off on merging this PR until that alias is up and working. I also confirmed with Sean that there is no way to differentiate between ultra and classic FWIW.

internal doc alias:
https://web.hypothes.is/help/bb-files

robertknight commented on lms/static/scripts/frontend_apps/components/test/LMSFilePicker-test.js:
```diff
@@ -339,4 +343,21 @@ describe('LMSFilePicker', () => {
+    it('renders no file message with a help link', async () => {
+      const wrapper = renderFilePicker();
+      await fakeApiCall.returnValues[0];
+      wrapper.update();
```
This `await ...; wrapper.update();` pair of statements follows the example of other tests, so I think it's OK to use it for the moment. This pattern is a little brittle though in that it will break if `LMSFilePicker` ever does any other additional async steps after the file list API call returns before it re-renders. A better solution is to wait for the `FileList` to be rendered in a non-loading state using the `waitForElement` helper:

```diff
diff --git a/lms/static/scripts/frontend_apps/components/test/LMSFilePicker-test.js b/lms/static/scripts/frontend_apps/components/test/LMSFilePicker-test.js
index 55f65b72..4218e475 100644
--- a/lms/static/scripts/frontend_apps/components/test/LMSFilePicker-test.js
+++ b/lms/static/scripts/frontend_apps/components/test/LMSFilePicker-test.js
@@ -7,6 +7,7 @@ import ErrorDisplay from '../ErrorDisplay';
 import FileList from '../FileList';
 import LMSFilePicker, { $imports } from '../LMSFilePicker';
 import mockImportedComponents from '../../../test-util/mock-imported-components';
+import { waitForElement } from '../../../test-util/wait';
 
 describe('LMSFilePicker', () => {
   // eslint-disable-next-line react/prop-types
@@ -351,9 +352,7 @@ describe('LMSFilePicker', () => {
 
     it('renders no file message with a help link', async () => {
       const wrapper = renderFilePicker();
-      await fakeApiCall.returnValues[0];
-      wrapper.update();
-      const fileList = wrapper.find('FileList');
+      const fileList = await waitForElement(wrapper, 'FileList');
       assert.equal(
         fileList.prop('noFilesMessage').props.href,
         'https://fake_help_link'
```

LMS007 replied:
I like waitForElement a lot better, much easier  as it takes care of those details which are hard to remember. I'll switch all the tests in this file which could use that call, perhaps in a second commit.

robertknight commented on lms/static/scripts/frontend_apps/components/LMSFilePicker.js:
```diff
@@ -14,6 +14,31 @@ import FileList from './FileList';
 
+/**
+ * @typedef NoFilesProps
+ * @prop {string} href - Helpful documentation url to link to
```
```suggestion
 * @prop {string} href - Helpful documentation URL to link to
```

robertknight commented on lms/static/scripts/frontend_apps/components/ContentSelector.js:
```diff
@@ -145,6 +148,9 @@ export default function ContentSelector({
           onCancel={cancelDialog}
           onSelectFile={selectBlackboardFile}
+          // An alias we maintain that provides multiple external documentation links for
+          // different versions of Blackboard (classic vs. ultra)
```
```suggestion
          // different versions of Blackboard (Classic vs. Ultra)
```

This is to make it clear that these terms are Blackboard's official names for these versions rather than just descriptions that we use.

robertknight reviewed (approved):
The code looks good. I had some minor suggestions regarding the name of the `NoFiles` component and a few other details. I had a query regarding the content that https://web.hypothes.is/help/bb-files points to. Currently this is a general page about using Hypothesis in Blackboard with no mention of Blackboard files. What is the plan regarding making this link point to something which is more specific to the issue here (eg. a page specifically about Blackboard Files) and is there an issue for that?

LMS007 commented:
> ral page about using Hypothesis in Blackboard with no mention of Blackboard files. What is the pla

https://web.hypothes.is/help/bb-files  page is just an alias that needs to point to a new content page which currently does not exist. I will hold off on merging this PR until that page is done. I'll follow up with support on a timeframe.
